core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,14 +201,20 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
       if (now - lastSeenMs > executorTimeoutMs) {
         logWarning(s"Removing executor $executorId with no recent heartbeats: " +
           s"${now - lastSeenMs} ms exceeds timeout $executorTimeoutMs ms")
-        scheduler.executorLost(executorId, SlaveLost("Executor heartbeat " +
-          s"timed out after ${now - lastSeenMs} ms"))
-          // Asynchronously kill the executor to avoid blocking the current thread
+        // Asynchronously kill the executor to avoid blocking the current thread
         killExecutorThread.submit(new Runnable {
           override def run(): Unit = Utils.tryLogNonFatalError {
             // Note: we want to get an executor back after expiring this one,
             // so do not simply call `sc.killExecutor` here (SPARK-8119)
             sc.killAndReplaceExecutor(executorId)
+            // In case of the executors which are not gracefully shut down, we should remove
```

Comment:
> In case of the executors which are not gracefully shut down

Do you mean `CoarseGrainedSchedulerBackend` is the only one that may shut down executors gracefully? And if it shuts down gracefully, is it OK to have 2 `RemoveExecutor` events?

Reply:
> Do you mean CoarseGrainedSchedulerBackend is the only one that may shut down executors non-gracefully? 

No. IIUC, it's executor's own behavior but not related to scheduler backend. An executor can be self-lost(no heartbeat) but still alive, e.g. due to network problem.

> And if it shuts down gracefully, is it OK to have 2 RemoveExecutor events?

Yes, we have protection for this case.

Reply:
then why do we only match `CoarseGrainedSchedulerBackend` here?

Reply:
Match `CoarseGrainedSchedulerBackend` covers all scheduler backend(Standalone, YARN, K8S, Mesos) that officially supported by Spark and we can guarantee that this issue can be resolved internally. But for other plugged scheduler backend, Spark seems has nothing to do with current `SchedulerBackend` interface at least for this issue.

Reply:
`SchedulerBackend` is private so we don't need to worry about plugins.

Let me ask the question again: is `CoarseGrainedSchedulerBackend` the only one that may shut down executors non-gracefully? According to what you said, it covers all the scheduler backend so the answer is yes?

Reply:
No, no.

Please pay attention to the comment above and note that, here, we're not saying that this issue happens due to `CoarseGrainedSchedulerBackend` may shut down executors non-gracefully. But we say that it may happens due to the executor itself is not gracefully shut down.

It's possible that `CoarseGrainedSchedulerBackend` asks someone executor to shutdown and the executor may shutdown  non-gracefully. And it's also possible that the host where the executor located suddenly crushed that makes the executor shutdown non-gracefully. 

Shutdown non-gracefully here only means that an executor shutdown but `CoarseGrainedSchedulerBackend` doesn't receive disconnect event from it. It really doesn't matter what causes the executor shutdown non-gracefully.

Reply:
do you mean only `CoarseGrainedSchedulerBackend` can't handle non-graceful executor shutdown well?

Reply:
Yeah, in this case, `CoarseGrainedSchedulerBackend` can't handle non-graceful executor shutdown well. But I am not sure what do you mean by "only", here? What other factors do you also care about?

Reply:
The executor may lost in network route, `CoarseGrainedSchedulerBackend` can't handle disconnect event, it happened in our production environment, too.

Reply:
only means we only need to match `CoarseGrainedSchedulerBackend` here, nothing else.

Reply:
Yeah, only.

Reply:
I agaree with this, I try to fix it here https://github.com/apache/spark/pull/26938, but it seems not very well, this patch is good to me.